### PR TITLE
Simplified withdrawal

### PR DESCRIPTION
**Pull request by dolgopolovwork**
We can simplify and make the withdrawal process more stable using brand new CAS functionality. Now just one node decides what UTXO to use.

Related PRs:
- Fixed genesis block in notary https://github.com/d3ledger/notary/pull/665
- Fixed genesis block in bootstrap https://github.com/d3ledger/bootstrap/pull/19
- Changelog https://github.com/d3ledger/deploy/pull/106

### CODE REVIEW

File: btc-withdrawal/src/main/kotlin/com/d3/btc/withdrawal/provider/UTXOProvider.kt [hunk 1]
```diff
@@ -26,257 +21,132 @@ import org.bitcoinj.core.TransactionOutput
...
     private val btcAddressStorage: BtcAddressStorage,
     private val usedUTXOProvider: UsedUTXOProvider
 ) {
 
+    private val utxoComparator: Comparator<TransactionOutput> = Comparator { output1, output2 ->
+        /*
+        Outputs are compared by values.
+        It will help us having a little amount of inputs.
+        Less inputs -> smaller tx size -> smaller fee*/
+        val valueComparison = -output1.value.value.compareTo(output2.value.value)
```
Comment (Mingela) [inline btc-withdrawal/src/main/kotlin/com/d3/btc/withdrawal/provider/UTXOProvider.kt:43]: isn't it a confusing comparison?

File: btc-withdrawal/src/main/kotlin/com/d3/btc/withdrawal/provider/UTXOProvider.kt [hunk 2]
```diff
@@ -26,257 +21,132 @@ import org.bitcoinj.core.TransactionOutput
...
                     unspent.parentTransactionDepthInBlocks >= confidenceLevel
+                    //We use registered clients outputs only
+                    && isAvailableOutput(unspent)
                     //Cannot use already used unspents
                     && !usedUTXOProvider.isUsed(withdrawalDetails, unspent).get()
-                    //We are able to use those UTXOs which height is not bigger then availableHeight
-                    && getUnspentHeight(unspent) <= availableHeight
-                    //We use registered clients outputs only
-                    && isAvailableOutput(availableAddresses, unspent)
+            if (availableToSpent) {
```
Comment (Mingela) [inline btc-withdrawal/src/main/kotlin/com/d3/btc/withdrawal/provider/UTXOProvider.kt]: availableToSpen**d** ?

File: btc-withdrawal/src/main/kotlin/com/d3/btc/withdrawal/provider/UTXOProvider.kt [hunk 3]
```diff
@@ -286,11 +156,10 @@ class UTXOProvider(
 
     // Checks if fee output was addressed to available address
     protected fun isAvailableOutput(
-        availableAddresses: Set<String>,
         output: TransactionOutput
     ): Boolean {
         val btcAddress = outPutToBase58Address(output)
-        return availableAddresses.contains(btcAddress)
+        return btcAddressStorage.isChangeAddress(btcAddress) || btcAddressStorage.isOurClient(btcAddress)
```
Comment (Mingela) [inline btc-withdrawal/src/main/kotlin/com/d3/btc/withdrawal/provider/UTXOProvider.kt]: I think I have seen this statement before 👀 Could we reuse other call if I'm correct?

File: btc-withdrawal/src/main/kotlin/com/d3/btc/withdrawal/provider/WithdrawalConsensusProvider.kt [hunk 4]
```diff
@@ -5,204 +5,162 @@
...
+    private fun handleConsensus(withdrawalDetails: WithdrawalDetails): Result<Unit, Exception> {
+        val utxo = ArrayList<SerializableUTXO>()
+        val consensusId = UUID.randomUUID().toString()
+        // Collect unspents
+        return bitcoinUTXOProvider.collectUnspents(withdrawalDetails, bitcoinConfig.confidenceLevel)
+            .flatMap { unspents ->
+                unspents.forEach { output ->
+                    val transaction = Transaction(btcNetworkConfigProvider.getConfig())
+                    val input = transaction.addInput(output)
+                    input.setParent(null)
```
Comment (Mingela) [inline btc-withdrawal/src/main/kotlin/com/d3/btc/withdrawal/provider/WithdrawalConsensusProvider.kt:72]: just to reduce memory consumption or what?

File: btc-withdrawal/src/main/kotlin/com/d3/btc/withdrawal/service/BtcRollbackService.kt [hunk 5]
```diff
@@ -41,8 +50,22 @@ class BtcRollbackService(
...
-
-        rollbackService.rollback(withdrawalFinalizationDetails, reason).fold(
+        withdrawalConsumer.getConsumerQuorum().flatMap { quorum ->
+            var transaction = rollbackService.createRollbackTransaction(withdrawalFinalizationDetails, reason, quorum)
+            if (btcTx != null) {
+                logger.info("Rollback Bitcoin transaction as well $btcTx")
+                val withdrawalRollbackData = WithdrawalRollbackData(
+                    withdrawalDetails,
+                    btcTx.inputs.map { input -> input.irohaKey() })
+                transaction = transaction.makeMutable()
```
Comment (Mingela) [inline btc-withdrawal/src/main/kotlin/com/d3/btc/withdrawal/service/BtcRollbackService.kt:60]: does `createRollbackTransaction` return a signed transaction?

File: btc-withdrawal/src/main/kotlin/com/d3/btc/withdrawal/service/WithdrawalTransferService.kt [hunk 6]
```diff
@@ -24,36 +22,36 @@ import org.springframework.stereotype.Component
...
+    fun withdraw(withdrawalConsensus: WithdrawalConsensus) {
+        var savedTx: Transaction? = null
+        transactionCreator.createTransaction(withdrawalConsensus).map { tx ->
+            savedTx = tx
+            registerWithdrawal(withdrawalConsensus.withdrawalDetails)
             registerTx(tx)
         }.failure { ex ->
-            btcRollbackService.rollback(withdrawalDetails, "Cannot create Bitcoin transaction")
+            val rollbackReason = "Cannot create Bitcoin transaction"
+            if (savedTx != null) {
```
Comment (Mingela) [inline btc-withdrawal/src/main/kotlin/com/d3/btc/withdrawal/service/WithdrawalTransferService.kt:45]: why not to handle this condition in `btcRollbackService.rollback` method?

File: btc-withdrawal/src/main/kotlin/com/d3/btc/withdrawal/transaction/TransactionsStorage.kt [hunk 7]
```diff
@@ -104,20 +88,22 @@ class TransactionsStorage(
...
+    val withdrawalConsensus: WithdrawalConsensus,
+    val txHex: String
+) {
     fun toJson() = gson.toJson(this)!!
 
     companion object {
         fun fromJson(json: String) = gson.fromJson(json, WithdrawalTransaction::class.java)!!
     }
-}
+}
```
Comment (Mingela) [inline btc-withdrawal/src/main/kotlin/com/d3/btc/withdrawal/transaction/TransactionsStorage.kt]: 😭

File: btc-withdrawal/src/main/kotlin/com/d3/btc/withdrawal/transaction/WithdrawalConsensus.kt [hunk 8]
```diff
@@ -5,42 +5,67 @@
...
 /**
  * Data class that holds all the information to decide how withdrawal transaction must be created
  */
-data class WithdrawalConsensus(val availableHeight: Int, val peers: Int) {
+data class WithdrawalConsensus(
+    val utxo: List<SerializableUTXO>,
+    val withdrawalDetails: WithdrawalDetails,
+    val id: String
+) {
+    fun toJson() = gson.toJson(this)!!
```
Comment (Mingela) [inline btc-withdrawal/src/main/kotlin/com/d3/btc/withdrawal/transaction/WithdrawalConsensus.kt:25]: move this to companion as well?

File: btc/src/main/kotlin/com/d3/btc/helper/input/TxInputExt.kt [hunk 9]
```diff
@@ -1,20 +1,17 @@
...
  */
-fun TransactionInput.irohaKey() = "${this.outpoint.index}_${this.outpoint.hash}".substring(0..32)
\ No newline at end of file
+fun TransactionInput.irohaKey() = "${this.outpoint.index}_${this.outpoint.hash}".substring(0..32)
+
+fun TransactionInput.verify(scriptBytes: ByteArray) {
+    val pubKey = Script(scriptBytes)
+    val myIndex = parentTransaction.inputs.indexOf(this)
+    scriptSig.correctlySpends(parentTransaction, myIndex.toLong(), pubKey)
+}
```
Comment (Mingela) [inline btc/src/main/kotlin/com/d3/btc/helper/input/TxInputExt.kt]: 😭

**Review by Mingela — APPROVED** [review]

File: btc-withdrawal/src/test/java/com/d3/btc/withdrawal/handler/NewSignatureEventHandlerTest.java [hunk 10]
```diff
@@ -149,7 +148,6 @@ public void testHandleGetSignaturesFail() {
         SetAccountDetailEvent event = new SetAccountDetailEvent(newSignatureDetail, signatureCollectorAccountId);
         newSignatureEventHandler.handle(event);
         verify(newSignatureEventHandler).broadcastIfEnoughSignatures(any(), any());
-        verify(btcRollbackService).rollback(any(), any());
-        verify(utxoProvider).unregisterUnspents(any(), any());
+        verify(btcRollbackService).rollback(any(), any(), any());
     }
 }
```
Comment (Alexey-N-Chernyshov) [inline btc-withdrawal/src/test/java/com/d3/btc/withdrawal/handler/NewSignatureEventHandlerTest.java]: eol

**Review by Alexey-N-Chernyshov — APPROVED** [review]

File: btc-withdrawal/src/main/kotlin/com/d3/btc/withdrawal/provider/UTXOProvider.kt (hunk 1 above)
Comment (dolgopolovwork) [reply btc-withdrawal/src/main/kotlin/com/d3/btc/withdrawal/provider/UTXOProvider.kt:43]: yes.

File: btc-withdrawal/src/main/kotlin/com/d3/btc/withdrawal/provider/UTXOProvider.kt (hunk 2 above)
Comment (dolgopolovwork) [reply btc-withdrawal/src/main/kotlin/com/d3/btc/withdrawal/provider/UTXOProvider.kt]: yup

File: btc-withdrawal/src/main/kotlin/com/d3/btc/withdrawal/provider/UTXOProvider.kt (hunk 3 above)
Comment (dolgopolovwork) [reply btc-withdrawal/src/main/kotlin/com/d3/btc/withdrawal/provider/UTXOProvider.kt]: ok

File: btc-withdrawal/src/main/kotlin/com/d3/btc/withdrawal/provider/WithdrawalConsensusProvider.kt (hunk 4 above)
Comment (dolgopolovwork) [reply btc-withdrawal/src/main/kotlin/com/d3/btc/withdrawal/provider/WithdrawalConsensusProvider.kt:72]: No. It's a trick. The constructor I want to use is not public and the only way to invoke it is by using `addInput` which returns an input. Unfortunately,  `addInput` has side effects. For instance, it sets `transaction` as an input parent transaction which is not needed. This is why I set it to null.

File: btc-withdrawal/src/main/kotlin/com/d3/btc/withdrawal/service/BtcRollbackService.kt (hunk 5 above)
Comment (dolgopolovwork) [reply btc-withdrawal/src/main/kotlin/com/d3/btc/withdrawal/service/BtcRollbackService.kt:60]: No. It's unsigned.

File: btc-withdrawal/src/main/kotlin/com/d3/btc/withdrawal/transaction/WithdrawalConsensus.kt (hunk 8 above)
Comment (dolgopolovwork) [reply btc-withdrawal/src/main/kotlin/com/d3/btc/withdrawal/transaction/WithdrawalConsensus.kt:25]: No. Look closely. It uses `this`.

Comment (Mingela) [reply btc-withdrawal/src/main/kotlin/com/d3/btc/withdrawal/transaction/WithdrawalConsensus.kt:25]: Then how about an extension function?

File: btc-withdrawal/src/main/kotlin/com/d3/btc/withdrawal/provider/UTXOProvider.kt (hunk 1 above)
Comment (Mingela) [reply btc-withdrawal/src/main/kotlin/com/d3/btc/withdrawal/provider/UTXOProvider.kt:43]: So it is better to refactor it 😆

**Pull request merged**